File: solution/206/answer.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def printAll(self, name):
        nodestr = "{name}: ".format(name=name);
        node = self
        while node:
            nodestr += "{val},".format(val=node.val)
            node = node.next
        print nodestr

class Solution(object):

    def reverseList(self, head):
        newhead = ListNode(0);
        while head:
            next = head.next;
            head.next = newhead.next;
            newhead.next = head;
            head = next;
        return newhead.next

if __name__ == "__main__":
    a = ListNode(1);
    b = ListNode(2);
    c = ListNode(3);
    d = ListNode(4);
    e = ListNode(5);
    a.next = b
    b.next = c
    c.next = d
    d.next = e
    head = a
    head.printAll("old");
    s = Solution()
    head = s.reverseList(head)
    head.printAll("new")

File: solution/206/answer.cpp
#include <stdio.h>

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL){}
};

void printList(ListNode* head, const char *name){
    printf("%s:", name);
    if(!head){
        printf("NULL\n");
        return;
    }
    while(head){
        printf("%d", head->val);
        head = head->next;
    }
    printf("\n");
}

class Solution{
public:
    ListNode* reverseList(ListNode* head){
        ListNode *newhead=NULL;
        while(head){
            ListNode *next = head->next;
            head->next = newhead;
            newhead = head;
            head = next;
        }
        return newhead;
    }
};

int main(){
    ListNode a(1);
    ListNode b(2);
    ListNode c(3);
    ListNode d(4);
    ListNode e(5);
    a.next = &b;
    b.next = &c;
    c.next = &d;
    d.next = &e;
    Solution s;
    ListNode *head = &a;
    printList(head, "old");
    head = s.reverseList(head);
    printList(head, "new");
    return 0;
}